mycroft/client/speech/silence.py

```diff
@@ -0,0 +1,402 @@
+
+    max_current_ratio_threshold: Optional[float] = None
+        Ratio of max/current energy below which audio is considered speech
+
+    current_energy_threshold: Optional[float] = None
+        Energy threshold above which audio is considered speech
+
+    silence_method: SilenceMethod = "vad_only"
+        Method for deciding if an audio chunk contains silence or speech
+    """
+
+    def __init__(
+        self,
+        sample_rate: int = 16000,
+        chunk_size: int = 960,
+        skip_seconds: float = 0,
+        min_seconds: float = 1,
+        max_seconds: typing.Optional[float] = 30,
+        speech_seconds: float = 0.3,
+        silence_seconds: float = 0.5,
+        before_seconds: float = 0.5,
+        max_energy: typing.Optional[float] = None,
+        max_current_ratio_threshold: typing.Optional[float] = None,
+        current_energy_threshold: typing.Optional[float] = None,
+        silence_method: SilenceMethod = SilenceMethod.VAD_ONLY,
+        plugin: str = None
+    ):
+        self.sample_rate = sample_rate
+        self.sample_width = 2  # 16-bit
+        self.sample_channels = 1  # mono
+        self.chunk_size = chunk_size
+        self.skip_seconds = skip_seconds
+        self.min_seconds = min_seconds
+        self.max_seconds = max_seconds
+        self.speech_seconds = speech_seconds
+        self.silence_seconds = silence_seconds
+        self.before_seconds = before_seconds
+
+        self.max_energy = max_energy
+        self.dynamic_max_energy = max_energy is None
+        self.max_current_ratio_threshold = max_current_ratio_threshold
+        self.current_energy_threshold = current_energy_threshold
+        self.silence_method = silence_method
+
+        # Verify settings
+        if self.silence_method in [
+            SilenceMethod.VAD_ONLY,
+            SilenceMethod.VAD_AND_RATIO,
+            SilenceMethod.VAD_AND_CURRENT,
+            SilenceMethod.ALL,
+        ]:
+            self.use_vad = True
+        else:
+            self.use_vad = False
+
+        if self.silence_method in [
+            SilenceMethod.VAD_AND_RATIO,
+            SilenceMethod.RATIO_ONLY,
+            SilenceMethod.ALL,
+        ]:
+            self.use_ratio = True
+            assert (
+                self.max_current_ratio_threshold is not None
+            ), "Max/current ratio threshold is required"
+        else:
+            self.use_ratio = False
+
+        if self.silence_method in [
+            SilenceMethod.VAD_AND_CURRENT,
+            SilenceMethod.CURRENT_ONLY,
+            SilenceMethod.ALL,
+        ]:
+            self.use_current = True
+            assert (
+                self.current_energy_threshold is not None
+            ), "Current energy threshold is required"
+        else:
+            self.use_current = False
+
+        # Voice detector
+        self.vad = plugin
+        if not plugin:
+            self.use_vad = False
+            if not self.use_current and not self.use_ratio:
```

Comment:
Consider:
```python
if not any((self.use_current, self.use_ratio)):
```

Makes it easier if we add other methods in the future (multiple VAD techniques?)

Reply:
new techniques will use plugins, this list should not be expanded

I personally find usage of any less readable unless theres a lot of entries, since this is just 2 conditions and its a port of someone else's code i will just leave it like it is